Using JAXP on XMLPrinter implementation

lcbarcellos opened the pull request:
Fixes #2749

The original implementation "manually" generates XML and does not take care of special cases, when illegal characters are present on attribute values. The proposed solution avoid the "manual" generation of XML, by using JAXP API for outputing XML structure.

Also, instead of generating the complete XML string in memory, user of XMLPrinter has now the option of streaming data directly to an output file, by using public methods of the class.

lcbarcellos committed:
Avoid use of temporary lists

lcbarcellos committed:
Fix: issue #2749 Generated XML is invalid

Replace "manual" generation of XML by use of
JAXP API. The test cases were replaced as well
so they now check for XML document equality
instead of string equality.

jlerbsc commented on javaparser-core-testing/src/test/java/com/github/javaparser/printer/XmlPrinterTest.java:
```diff
@@ -25,38 +25,73 @@
+        }
+    }
+
+    public Document getDocument(String xml) throws SAXException, IOException {
```
Is there no need to close the inputStream?

lcbarcellos replied:
As per [JavaDoc](https://docs.oracle.com/javase/8/docs/api/java/io/ByteArrayInputStream.html#close--), closing a ByteArrayInputStream has no effect.

jlerbsc commented on javaparser-core/src/main/java/com/github/javaparser/printer/XmlPrinter.java:
```diff
@@ -42,76 +47,120 @@ public XmlPrinter(boolean outputNodeType) {
+    }
+
+    public StringWriter stringWriterOutput(Node node, String name) {
+        StringWriter stringWriter = new StringWriter();
```
I imagine that this method allows you to implement data streaming in a file, perhaps you can specify in javadoc that the resource returned by the method must be closed by the user?

lcbarcellos committed:
Some refactoring about checked exceptions

lcbarcellos committed:
Additional assertions on outputNode method parameters

lcbarcellos committed:
Code documentation

lcbarcellos committed:
Renaming and refactoring (no functional change)

lcbarcellos committed:
Richer error reporting on test

lcbarcellos committed:
Test cleanup and temp file support

lcbarcellos committed:
Test streaming to file

lcbarcellos committed:
Test custom XML

lcbarcellos committed:
Java 8 compatibility fix.

No need to worry about which charset is used, as the same
is going to be used for saving and loading.

jlerbsc committed:
Merge branch 'master' into master
